### PR TITLE
Extract DeploymentContext from InternalDeploymentTester

### PR DESCRIPTION
**Pull request by mpolden**
`MetricsReporterTest` should give you an idea of the API.

### CODE REVIEW

File: controller-server/src/test/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentContext.java [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+        this.applicationId = TenantAndApplicationId.from(instanceId);
+        this.instanceId = instanceId;
+        this.testerId = TesterId.of(instanceId);
+        this.jobs = tester.controller().jobController();
+        this.routing = tester.serviceRegistry().routingGeneratorMock();
+        this.cloud = tester.serviceRegistry().testerCloud();
+        this.runner = runner;
+        this.tester = tester;
+        this.readyJobsTrigger = new ReadyJobsTrigger(tester.controller(), maintenanceInterval, jobControl);
+        this.nameServiceDispatcher = new NameServiceDispatcher(tester.controller(), maintenanceInterval, jobControl,
```
Comment (mpolden) [inline controller-server/src/test/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentContext.java]: Things that can be safely shared across contexts should probably continue living in `InternalDeploymentTester`, so these maintainers can be moved back or perhaps to `ControllerTester`.

Comment (jonmv) [reply controller-server/src/test/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentContext.java]: Yes.

File: controller-server/src/test/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentContext.java [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+    }
+
+    /** Returns the last submitted application version */
+    public Optional<ApplicationVersion> lastSubmission() {
+        return Optional.ofNullable(lastSubmission);
+    }
+
+    /** Runs and returns all remaining jobs for the application, at most once, and asserts the current change is rolled out. */
+    public DeploymentContext completeRollout() {
+        readyJobsTrigger.run();
```
Comment (jonmv) [inline controller-server/src/test/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentContext.java]:
```suggestion
        triggerJobs();
```

File: controller-server/src/test/java/com/yahoo/vespa/hosted/controller/deployment/InternalDeploymentTester.java [hunk 3]
```diff
@@ -137,26 +105,41 @@ public InternalDeploymentTester() {
...
+        return nameServiceDispatcher;
+    }
+
+    /** Returns the default deployment context owned by this */
+    public DeploymentContext deploymentContext() {
+        return defaultContext;
+    }
+
+    /** Create a new deployment context for given application */
+    public DeploymentContext newDeploymentContext(String tenantName, String applicationName, String instanceName) {
```
Comment (jonmv) [inline controller-server/src/test/java/com/yahoo/vespa/hosted/controller/deployment/InternalDeploymentTester.java]: It's possible we should leave the instance out here, and support it in the context somehow. Let's see how it turns out.

File: controller-server/src/test/java/com/yahoo/vespa/hosted/controller/deployment/InternalDeploymentTester.java [hunk 4]
```diff
@@ -137,26 +105,41 @@ public InternalDeploymentTester() {
...
-                                                           instanceName,
-                                                           nextProjectId.getAndIncrement());
+        return newDeploymentContext(tenantName, applicationName, instanceName).application();
     }
 
     /** Submits a new application, and returns the version of the new submission. */
-    public ApplicationVersion newSubmission(TenantAndApplicationId id, ApplicationPackage applicationPackage,
-                                            SourceRevision revision, String authorEmail, long projectId) {
-        return jobs.submit(id, revision, authorEmail, projectId, applicationPackage, new byte[0]);
+    public ApplicationVersion newSubmission(TenantAndApplicationId id, ApplicationPackage applicationPackage, SourceRevision sourceRevision) {
```
Comment (jonmv) [inline controller-server/src/test/java/com/yahoo/vespa/hosted/controller/deployment/InternalDeploymentTester.java]: We should probably inline methods like this? (Work for later.)

**Review by jonmv — APPROVED** [review]
Nice!

File: controller-server/src/test/java/com/yahoo/vespa/hosted/controller/deployment/InternalDeploymentTester.java (hunk 4 above)
Comment (mpolden) [reply controller-server/src/test/java/com/yahoo/vespa/hosted/controller/deployment/InternalDeploymentTester.java]: I think these methods should be removed altogether, and rather have tests always use a context.

Comment (jonmv) [reply controller-server/src/test/java/com/yahoo/vespa/hosted/controller/deployment/InternalDeploymentTester.java]: Exactly.

**Pull request merged**